### ISSUE TITLE
New ConfigureOptions APIs that use TryAddEnumerable

### ISSUE THREAD

**Issue by weichch** [dotnet/runtime#42358]
## Background and Motivation

`ConfigureOptions()` method is often used by library authors who need to add their own types to configure options. 

Currently the method uses `AddTransient` internally when adding each of the `IConfigureOptions<>`, `IPostConfigureOptions<>`, and `IValidateOptions<>` found. This often can lead to a problem where if the `ConfigureOptions()` method gets called more than once, the configure services will be registered more than once, therefore the options type will be configured more than once when resolving `IOptions<T>.Value`.

The current workaround is adding types manually using `TryAddEnumerable` method against each interface types instead of `ConfigureOptions`. It would be great if we could have new APIs equivalent  to `ConfigureOptions()` but uses `TryAddEnumerable` internally, so that the configure method in custom libraries could look a lot nicer.

## Proposed API

```csharp
public static IServiceCollection TryConfigureOptions<T>(this IServiceCollection services)
    where T : class
{
    return services.TryConfigureOptions(typeof(T));
}

public static IServiceCollection TryConfigureOptions<T>(this IServiceCollection services, Type configureType)
    where T : class
{
    services.AddOptions();
    IEnumerable<Type> enumerable = FindConfigurationServices(configureType);
    foreach (Type item in enumerable)
    {
        // Use TryAddEnumerable here
        services.TryAddEnumerable(ServiceDescriptor.Describe(
            item,
            configureType,
            ServiceLifetime.Transient));
    }
    return services;
}
```

## Usage Examples

```csharp
services.TryConfigureOptions<ConfigureMyOptions>();
services.TryConfigureOptions(typeof(ConfigureMyOptions));

public class ConfigureMyOptions 
    : IConfigureOptions<MyOptions>,
    IPostConfigureOptions<MyOptions>,
    IValidateOptions<MyOptions>
{
    public void Configure(MyOptions options)
    {
    }

    public void PostConfigure(string name, MyOptions options)
    {
    }

    public ValidateOptionsResult Validate(string name, MyOptions options)
    {
        return ValidateOptionsResult.Success;
    }
}
```

**Comment by ericstj** [comment]
@maryamariyan @ViktorHofer @haok @davidfowl for discussion

**Comment by HaoK** [comment]
Just to make sure I understand the scenario, since libraries might end up having `services.ConfigureOptions<T>` called more than once on the same T, the configurations in T may run multiple times.  And this new TryConfigureOptions is what libraries should to ensure that their options setup only runs once.  That seems reasonable to me if I understand the intent correctly...

**Comment by weichch** [comment]
@HaoK Yes, that's what I'm trying to propose. Even in some Microsoft libraries `TryAddEnumerable` is used to avoid running configuration for T more than once, like this [example](https://github.com/dotnet/aspnetcore/blob/master/src/Azure/AzureAD/Authentication.AzureAD.UI/src/AzureADAuthenticationBuilderExtensions.cs#L123).

**Comment by mburumaxwell** [comment]
Any hopes on this?
@HaoK is this an area for which you accept contributions?

**Comment by bart-vmware** [comment]
Running into this as well, a fix would be greatly appreciated.
